### PR TITLE
Feature/prisma test setup

### PR DESCRIPTION
**Pull request by eadsoy**
Looks like the test script is picking up the correct db but the migration is failing 
Screen Shot 2022-03-01 at 15 35 56

### CODE REVIEW

**Comment by eadsoy** [comment]
> @eadsoy I think either the code here is wrong, or the comment is wrong, or both. I think nanoseconds are seconds * 1B.

Yes, the comment was wrong. I corrected it in the next commit e705599, https://github.com/NEAR-Edu/near-certification-tools/blob/e7055994ce471ba953008416fafed5d10e743f1b/web-app/helpers/expiration-date.ts#L77

Previously `issued_at` was coming in as a date string, now it's again a string but a UNIX timestamp in  milliseconds. It's already in the format we want, so there is no need to convert it. That is why I removed the dayjs conversion.

File: web-app/helpers/expiration-date.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      {
+        moment: '2022-04-07T16:25:59+00:00', // Most recent activity
+        diff_to_next_activity: null // null, since no >180-day period was found
+      }
+    ]
+   * 
+   * BOTH queries are set up to return the desired date in a column called 'moment'.
+   */
+
+  // *issue_date* <-----------Query - 1-----------> *last_activtiy*  <-----------Query - 2 -----------> *now (render date)*
```
Comment (sccheruku) [inline web-app/helpers/expiration-date.ts:70]: I'm hoping my comments do not add to confusion. Some clarification would be enough. 

I can't tell just by looking at the query if the following scenario is being covered? 

`ID-2022-01-01` <---- `Activity-2022-04-01` <--- `gt 6m` ---> `Activity-2022-10-15` <---  `render-2022-12-01`

The certification should be expired on 2022-10-01 (6 months have passed since the last activity on 2022-04-01)

**Review by sccheruku — APPROVED** [review]

**Comment by sccheruku** [comment]
Approved. But I am wondering if we tried to use Prisma ORM code instead of using the raw sql approach ? 
One downside of using raw sql that it opens us to sql injection attacks. It's not a big deal in this case, because we are working with a read only database. 

I'll share a gist to an example I was working on using Prisma ORM code, but I do not think we should block this PR. We have 6 months to decide if we want to refactor anything, right? 

https://www.loom.com/share/d87f4807646c4b79b2f5eae3ab4e2879

Here is the gist: https://gist.github.com/sccheruku/94f0d1406c3ebace3de4045c095eef9f (on testnet, my query took about 300 seconds when I set the minimum activity period to 7 days instead of 180 days)

It's a first iteration of the code, so it does not tackle all the edge cases @eadsoy  covered in the tests.

File: web-app/helpers/expiration-date.ts (hunk 1 above)
Comment (sccheruku) [reply web-app/helpers/expiration-date.ts:70]: Nevermind, I think you have this test case to cover the scenario above.  https://github.com/NEAR-Edu/near-certification-tools/blob/feature/prisma-test-setup/web-app/test/__tests__/expiration-date.test.ts#L95

**Comment by ryancwalsh** [comment]
@sccheruku Thanks for the review, the gist, and the video. I've used Prisma before but have definitely run into limitations, and I'm not 100% sure it can handle what we're trying to do.

If we _can_ use Prisma, then I agree that we should seriously consider it. Abstracting away from raw SQL can provide some flexibility (allows us to swap out the database underneath, theoretically).

I might be misunderstanding your approach, but something doesn't feel ideal about iterating an indeterminately large number of times to check for each period (currently defined as 6 months, but that's an arbitrary number that might change int he future too).

We can talk about it live today.
